cirq/circuits/circuit.py

```diff
@@ -1387,6 +1390,126 @@ def tetris_concat(
 
         return cirq.Circuit(buffer[offset : offset + n_acc])
 
+    def get_independent_qubit_sets(self) -> List[Set['cirq.Qid']]:
+        """Divide circuit's qubits into independent qubit sets.
+
+        Independent qubit sets are the qubit sets such that there are
+        no entangling gates between qubits belonging to different sets.
+        If this is not possible, a sequence with a single factor (the whole set of
+        circuit's qubits) is returned.
+
+        >>> q0, q1, q2 = cirq.LineQubit.range(3)
+        >>> circuit = cirq.Circuit()
+        >>> circuit.append(cirq.Moment(cirq.H(q2)))
+        >>> circuit.append(cirq.Moment(cirq.CZ(q0,q1)))
+        >>> circuit.append(cirq.H(q0))
+        >>> print(circuit)
+        0: ───────@───H───
+                  │
+        1: ───────@───────
+        <BLANKLINE>
+        2: ───H───────────
+        >>> [sorted(qs) for qs in circuit.get_independent_qubit_sets()]
+        [[cirq.LineQubit(0), cirq.LineQubit(1)], [cirq.LineQubit(2)]]
+
+        Returns:
+            The list of independent qubit sets.
+
+        """
+        uf = networkx.utils.UnionFind(self.all_qubits())
+        for op in self.all_operations():
+            if len(op.qubits) > 1:
+                uf.union(*op.qubits)
+        return sorted([qs for qs in uf.to_sets()], key=min)
+
+    def factorize(self: CIRCUIT_TYPE) -> Iterable[CIRCUIT_TYPE]:
+        """Factorize circuit into a sequence of independent circuits (factors).
+
+        Factorization is possible when the circuit's qubits can be divided
+        into two or more independent qubit sets. Preserves the moments from
+        the original circuit.
+        If this is not possible, returns the set consisting of the single
+        circuit (this one).
+
+        >>> q0, q1, q2 = cirq.LineQubit.range(3)
+        >>> circuit = cirq.Circuit()
+        >>> circuit.append(cirq.Moment(cirq.H(q2)))
+        >>> circuit.append(cirq.Moment(cirq.CZ(q0,q1)))
+        >>> circuit.append(cirq.H(q0))
+        >>> print(circuit)
+        0: ───────@───H───
+                  │
+        1: ───────@───────
+        <BLANKLINE>
+        2: ───H───────────
+        >>> for i, f in enumerate(circuit.factorize()):
+        ...     print("Factor {}".format(i))
+        ...     print(f)
+        ...
+        Factor 0
+        0: ───────@───H───
+                  │
+        1: ───────@───────
+        Factor 1
+        2: ───H───────────
+
+        Returns:
+            The sequence of circuits, each including only the qubits from one
+            independent qubit set.
+
+        """
+
+        qubit_factors = self.get_independent_qubit_sets()
+        if len(qubit_factors) == 1:
+            return (self,)
+        return (
+            self._with_sliced_moments([m[qubits] for m in self.moments]) for qubits in qubit_factors
```

Comment:
I guess I originally had the idea that

```python
self._with_sliced_moments([m[qubits] for m in self.moments])
```

could be the implementation of `slice_by_qubits`. E.g. if you already know your circuit is composed of independent factors. It would raise an exception otherwise. The `factorize` function can use it.

However, the semantics of `Moment.__getitem__` will not raise an exception. Rather, it will return all operations on which any of the qubits operate. We should probably add a comment to this line explaining the semantics of `m[qubits]`, namely that since we've already found the independent qubit sets we're guaranteed to get moments that *only* operate on the given qubits.

Reply:
I see, thanks for the explanations. Will remove `slice_by_qubits` method.